lib/agent.js

```diff
@@ -296,8 +336,13 @@ Agent.prototype._nextMessageId = function nextToken() {
   return this._lastMessageId
 }
 
+/**
+ * Entry point for a new client-side request.
+ * @param {*} url A String representing a CoAP URL, or an object with the appropriate parameters. 
+ */
 Agent.prototype.request = function request(url) {
   this._init()
```

Comment:
hm ... Agent._init() is setting _closing to false directly as first action https://github.com/mcollina/node-coap/blob/master/lib/agent.js#L60 ... so I do not really get the difference

Reply:
Hmm good catch, I commented it out and tested it against the unit tests and actually with receiving singular and block 1 transfers, and didn't see any issues.  That was in there from @Jamezo97 's [PR](https://github.com/mcollina/node-coap/pull/178/files?file-filters%5B%5D=.js#diff-d8bb2c7932a22406705bd0c42d2cb3d720574c245ba2cc81993a7d8d563b0e3eR338)

I don't see any reason to keep that in there, I can go ahead and remove it.

Reply:
Please remove it

Reply:
I just realized why that flag was in there

It seems that in rare circumstances, the CoAP agent is flagged to be closed, and whilst in the process of closing the DGRAM socket (between '_cleanUp' and '_doClose') there exists a small window where another CoAP request can be made.
Even if this request is made, the socket is still closed, and the response to the request is never received.
Resetting the '_closing' flag when a new request is made seems to solve the issue.

Reply:
I still do not get it :-)
When a request is done the first call goes to init() and the first what "Init" does is "_closing = false" and all other code there is kind of synchronous

BTW ... should in https://github.com/mcollina/node-coap/blob/7a4b770f7fd92f389b7bde3d1c57b6d1822882b0/lib/agent.js#L388 be missing the "if this._closing" check? Could that be the reason?

All ways to doClose are checked that closing is set - beside this one above ...

Reply:
If your words are correct the on call of init _closing gets set to false and the init does nothing else because the socket is still there ... and theny you reset it again ...so basically I do'nt get it else ... code flow wise.

Reply:
The missing check for `that._closing` was indeed causing problems (in my case when performing receiving block2 transfers when a multicast request was made). I included it in #237. So I would assume the setting of the `this._closing` flag could be removed now. Maybe you can do some more testing with the latest version of `master`, @ats-org?

Reply:
@JKRhb I've actually haven't come across the premature closing issue I pulled that over from #17, I just reviewed your change and @Apollon77 recommendation and agree can go ahead and be removed.

Reply:
@JKRhb @Apollon77,

I've remerged in master to catch it up again. I've run the unit tests and everything is still passing, I've tested with my IOT devices and some dummy data locally and everything appears to be working a-ok.

Reply:
Great, thank you for continuing to work on this PR! Maybe you could add a few more test cases so that the coverage stays more or less the same?

Reply:
It's been a few years since I've worked with CoAP so the details behind the change are quite fuzzy. As I wrote in the original PR, there was some async code which acted upon that _closing flag. There was a short window of time where the agent would be flagged for closing, but a new request was received. The _closing flag wasn't reset, and so the socket was closed and a response couldn't get out.

I believe I was stress testing the code at that point, so there were many simultaneous connections, and the event handler in the NodeJS engine was might have taken slightly longer to service immediate callbacks etc which would have made this issue a bit more noticeable.

Note that the commit for this line of code was done on the 7th of May, 2018
https://github.com/mcollina/node-coap/pull/233/commits/e741958b3418fe1953e084fe11443d7fea872e89

And the `this._closing = false;` was added to the Agent._init function on the [2nd of March, 2019](https://github.com/mcollina/node-coap/blame/2de089a17b43c66b4a8a605366a9bd40cd6fefdc/lib/agent.js#L60)

If you removed that line from Agent._init and stress test the system it might rear its head again, not too sure  ¯\_(ツ)_/¯

So anyway, as you've all agreed, it's no longer required as it's now covered in the init function

Reply:
@Jamezo97 Thank you for clarifying the origins of the line and the confirmation that it can be removed :)